### PR TITLE
[query/pcode] addLiteral returns PValue, literalMap stores PSettable

### PR DESCRIPTION
**Pull request by chrisvittal**
WIP. Need to figure out addLiteral for DependentEmitFunction.

### CODE REVIEW

File: hail/src/main/scala/is/hail/asm4s/CodeBuilder.scala [hunk 1]
```diff
@@ -33,13 +33,9 @@ trait CodeBuilderLike {
 
   def result(): Code[Unit]
 
-  val localBuilder: SettableBuilder = new SettableBuilder {
-    def newSettable[T](name: String)(implicit tti: TypeInfo[T]): Settable[T] = mb.newLocal[T](name)
-  }
+  def localBuilder: SettableBuilder = mb.localBuilder
```
Comment (cseed) [inline hail/src/main/scala/is/hail/asm4s/CodeBuilder.scala:36]: Just a drive-by remark: I kind of regret choosing the name `localBuilder`.  I think it should just be `newLocal _`, where newLocal matches the exsiting local creation functions we have.  Then SettableBuilder[T] is just a type alias (if it needs exists at all) for `(String) => (TypeInfo[T]) => Settable[T]`.

Think about it.

File: hail/src/main/scala/is/hail/expr/ir/Emit.scala [hunk 2]
```diff
@@ -611,9 +611,9 @@ private class Emit[C](
       case F64(x) =>
         present(pt, const(x))
       case s@Str(x) =>
-        present(pt, mb.addLiteral(x, coerce[PString](s.pType)))
+        present(mb.addLiteral(x, coerce[PString](s.pType)).get)
```
Comment (cseed) [inline hail/src/main/scala/is/hail/expr/ir/Emit.scala:614]: `.get` shouldn't be necessary here.  There should be an implicit conversion from PValue to PCode.  If there's not, can we add it?

File: hail/src/main/scala/is/hail/asm4s/CodeBuilder.scala (hunk 1 above)
Comment (chrisvittal) [reply hail/src/main/scala/is/hail/asm4s/CodeBuilder.scala:36]: I'm not sure what the type signature of the lambda would be in order to write something like
```scala
// Somewhere
type SettableBuilder = (String) => TypeInfo[???] => Settable[???] // we need the callers of this to choose ???
// EmitFunctionBuilder.scala
def newPSettable(_pt: PType, mkSettable: SettableBuilder, baseName: String): PSettable
// PCanonicalArray.scala
Object PCanonicalIndexableSettable {
  def apply(mkSettable: SettableBuilder, pt: PContainer, name: String): PCanonicalIndexableSettable = {
    new PCanonicalIndexableSettable(pt,
      mkSettable(s"${ name }_a")(LongInfo),
      mkSettable(s"${ name }_length")(IntInfo),
      mkSettable(s"${ name }_elems_addr")(LongInfo))
  }
}
```
I'm more than willing to do this. It feels like the right thing to do, but I'm not sure how to actually write the scala code such that it type checks.

Comment (chrisvittal) [reply hail/src/main/scala/is/hail/asm4s/CodeBuilder.scala:36]: I'm not sure this can work. The type that scala infers for `newLocal _` is `(string) => LocalRef[Nothing]`.

File: hail/src/main/scala/is/hail/expr/ir/Emit.scala (hunk 2 above)
Comment (chrisvittal) [reply hail/src/main/scala/is/hail/expr/ir/Emit.scala]: Done

File: hail/src/main/scala/is/hail/asm4s/CodeBuilder.scala (hunk 1 above)
Comment (chrisvittal) [reply hail/src/main/scala/is/hail/asm4s/CodeBuilder.scala:36]: I still haven't figured this out, so I'm continuing on with `SettableBuilder`

File: hail/src/main/scala/is/hail/expr/ir/EmitFunctionBuilder.scala [hunk 3]
```diff
@@ -1044,16 +1045,30 @@ class DependentEmitFunctionBuilder[F](
...
 
+  def newDepPField(pc: PCode): PValue = {
+    val ti = typeToTypeInfo(pc.pt)
+    val field = newPField(pc.pt)
+    dep_apply_method.setFields += { (obj: lir.ValueX) =>
+      val code = pc.code
+      // XXX below assumes that the first settable is the 'base' of the PSettable
+      val baseField = field.settableTuple()(0).asInstanceOf[ThisFieldRef[_]]
+      code.end.append(lir.putField(className, baseField.name, ti, obj, code.v))
+      // FIXME need to initialize other potential settables in the PSettable here
```
Comment (chrisvittal) [inline hail/src/main/scala/is/hail/expr/ir/EmitFunctionBuilder.scala:1064]: This is still definitely the part I feel least confident in. I'm not sure this code is ever used, so it continues to pass tests, but I would feel much better if I could store everything properly.

**Comment by chrisvittal** [comment]
@cseed bump for second review

**Comment by chrisvittal** [comment]
@cseed bump

File: hail/src/main/scala/is/hail/expr/ir/EmitFunctionBuilder.scala (hunk 3 above)
Comment (chrisvittal) [reply hail/src/main/scala/is/hail/expr/ir/EmitFunctionBuilder.scala:1064]: @cseed This is the only bit I where I specifically need feedback from you, in dependent methods.

Comment (tpoterba) [reply hail/src/main/scala/is/hail/expr/ir/EmitFunctionBuilder.scala:1064]: yeah, I agree this isn't ideal. I think we'll break this when we extend PCode, but it seems fine for now.
